panoramapublic/src/org/labkey/panoramapublic/query/speclib/SpecLibInfoTableInfo.java

```diff
@@ -0,0 +1,199 @@
+import org.labkey.api.data.SQLFragment;
+import org.labkey.api.data.Table;
+import org.labkey.api.data.TableInfo;
+import org.labkey.api.data.TableSelector;
+import org.labkey.api.query.LookupForeignKey;
+import org.labkey.api.query.QuerySettings;
+import org.labkey.api.query.QueryUpdateService;
+import org.labkey.api.query.QueryView;
+import org.labkey.api.query.RowIdQueryUpdateService;
+import org.labkey.api.query.UserSchema;
+import org.labkey.api.query.ValidationException;
+import org.labkey.api.security.User;
+import org.labkey.api.security.UserPrincipal;
+import org.labkey.api.security.permissions.Permission;
+import org.labkey.api.util.HtmlString;
+import org.labkey.panoramapublic.PanoramaPublicManager;
+import org.labkey.panoramapublic.PanoramaPublicSchema;
+import org.labkey.panoramapublic.model.speclib.SpecLibInfo;
+import org.labkey.panoramapublic.query.PanoramaPublicTable;
+import org.springframework.validation.Errors;
+
+import java.io.IOException;
+import java.io.Writer;
+
+public class SpecLibInfoTableInfo extends PanoramaPublicTable
+{
+    public SpecLibInfoTableInfo(PanoramaPublicSchema schema, ContainerFilter cf)
+    {
+        super(PanoramaPublicManager.getTableInfoSpecLibInfo(), schema, cf, getJoinSql(), new SQLFragment(" exp.Container "));
+
+        var dependencyTypeCol = getMutableColumn("DependencyType");
+        if (dependencyTypeCol != null)
+        {
+            dependencyTypeCol.setFk(new LookupForeignKey()
+            {
+                @Override
+                public TableInfo getLookupTableInfo()
+                {
+                    return getUserSchema().getTable(PanoramaPublicSchema.TABLE_LIB_DEPENDENCY_TYPE, cf);
+                }
+            });
+        }
+
+        var sourceTypeCol = getMutableColumn("SourceType");
+        if (sourceTypeCol != null)
+        {
+            sourceTypeCol.setFk(new LookupForeignKey()
+            {
+                @Override
+                public TableInfo getLookupTableInfo()
+                {
+                    return getUserSchema().getTable(PanoramaPublicSchema.TABLE_LIB_SOURCE_TYPE, cf);
+                }
+            });
+        }
+
+        var sourcePasswordCol = getMutableColumn("SourcePassword");
+        if (sourcePasswordCol != null)
+        {
+            sourcePasswordCol.setDisplayColumnFactory(new PasswordDisplayColumnFactory());
+        }
+    }
+
+    @Override
+    public boolean hasPermission(@NotNull UserPrincipal user, @NotNull Class<? extends Permission> perm)
+    {
+        return getContainer().hasPermission(user, perm);
+    }
+
+    @Override
+    public QueryUpdateService getUpdateService()
+    {
+        return new SpecLibInfoQueryUpdateService(this);
+    }
+
+    private static SQLFragment getJoinSql()
+    {
+        SQLFragment joinToExpAnnotSql = new SQLFragment(" INNER JOIN ");
+        joinToExpAnnotSql.append(PanoramaPublicManager.getTableInfoExperimentAnnotations(), "exp");
```

Comment:
I assume this "exp" is why the constructor SQLFragment uses "exp.Container"? If so, perhaps a constant or something else to link them more strongly?

Reply:
Done.